### PR TITLE
Add a Testing Policy

### PR DESCRIPTION
**Pull request by mattcaswell**

### CODE REVIEW

File: policies/testing.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+Testing Policy
+==============
+
+This applies to all stable and development branches of the main code repository.
+Except where noted below:
+
+- All Pull Requests adding new functionality to the applications, libraries,
+providers or engines must include suitable tests.
+- All Pull Requests fixing a functional behavior defect in the applications,
+libraries, providers or engines must include a test for that defect.
```
Comment (kroeckx) [inline policies/testing.md:10]: Do have or need a definition functional behaviour? Is any bugfix a functional behaviour change?

**Comment by kroeckx** [comment]
This policy should probably include adding fuzzers.

File: policies/testing.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+libraries, providers or engines do not require tests to be added. For example
+the following types of changes do not require tests:
+ - Changes to comments, formatting, internal naming or similar
+ - Fixes for performance
+ - Refactoring
+
+To be explicit the above statement means that tests are not required for changes
+in:
+ - Documentation (including CHANGES.md/NEWS.md)
+ - The test suite
```
Comment (levitte) [inline policies/testing.md:22]: Hmmm...  we have a test utilities library, and we *do* have a test for it: `test/recipes/01-test_test.t` and `test/test_test.c`...

Comment (paulidale) [reply policies/testing.md:22]: These tests haven't aged well.  There have been a lot of things added to the test framework which aren't covered.

**Comment by mattcaswell** [comment]
- Include wording about difficult to reproduce error conditions
- Tests could be added in different PRs
- Fuzz testing for substantial features

File: policies/testing.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+- All Pull Requests adding new functionality to the applications, libraries,
+providers or engines must include suitable tests.
+- All Pull Requests fixing a functional behavior defect in the applications,
+libraries, providers or engines must include a test for that defect.
+
+Pull Requests that do not change the functional behaviour of the applications,
+libraries, providers or engines do not require tests to be added. For example
+the following types of changes do not require tests:
+ - Changes to comments, formatting, internal naming or similar
+ - Fixes for performance
```
Comment (kroeckx) [inline policies/testing.md:16]: If it's about performance, and it was not something we monitored under the performance testing before, should we not add it to that? Clearly someone thought it was important enough to fix it so that we want to monitor it for regressions.

File: policies/testing.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ - Refactoring
```
Comment (kroeckx) [inline policies/testing.md:17]: Refactoring seems to be a good time to check that corner cases are tested. It might not need new tests, but it seems to be a good time to check that there are enough tests.

File: policies/testing.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+Performance Testing
+-------------------
+
+Performance testing should be performed automatically via CI on a regular basis
+for certain defined components.
+
+Components that should be performance tested via CI include:
+ - openssl speed for all supported algorithms
+ - openssl s_time
```
Comment (kroeckx) [inline policies/testing.md:38]: This seems to mention implementation details rather than what we want to do. I prefer that the policy juts mentions what kind of things should be tested.

File: policies/testing.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+ - Build system
+ - Demos
+
+Pull Requests that only affect minor cosmetic output do not require tests
+
+Performance Testing
+-------------------
+
+Performance testing should be performed automatically via CI on a regular basis
+for certain defined components.
```
Comment (kroeckx) [inline policies/testing.md:34]: It says on regular basis. Does that mean that it's not run for each pull request?

File: policies/testing.md (hunk 1 above)
Comment (mattcaswell) [reply policies/testing.md:10]: I've added a paragraph attempting to explain what is meant by this.

File: policies/testing.md (hunk 2 above)
Comment (mattcaswell) [reply policies/testing.md:22]: The test policy says that tests are not required....but it doesn't prevent us from adding them if we want them anyway.

File: policies/testing.md (hunk 3 above)
Comment (mattcaswell) [reply policies/testing.md:16]: I've added a non-mandatory "Recommendations" section and added recommendation to consider whether a performance test is appropriate.

File: policies/testing.md (hunk 4 above)
Comment (mattcaswell) [reply policies/testing.md:17]: I've added a recommendation to this effect.

File: policies/testing.md (hunk 5 above)
Comment (mattcaswell) [reply policies/testing.md:38]: I've reworded this and dropped specific mention of speed and s_time

File: policies/testing.md (hunk 6 above)
Comment (mattcaswell) [reply policies/testing.md:34]: I wasn't imagining this being run on every pull request.

**Comment by mattcaswell** [comment]
> Include wording about difficult to reproduce error conditions
Tests could be added in different PRs
Fuzz testing for substantial features

Fixup pushed addressing the above items as well as the feedback so far.

File: policies/testing.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+Within this policy "functional behaviour" means what the system does, i.e.
+given a set of inputs it will produce a set of outputs. This does not include
+how the system does it. For example refactoring or performance improvements do
+not affect functional behaviour.
+
+Except where noted below:
+
+- All Pull Requests adding new functionality to the applications, libraries,
+providers or engines must include suitable tests.
+- All Pull Requests fixing a functional behavior defect in the applications,
```
Comment (t8m) [inline policies/testing.md:15]: I think it is overly strict to require a test for every small bug fix. Especially for the cases where writing the test would require adding significant infrastructure (tests depending on triggering a malloc failure, tests having to add a new provider implementation or other cases). IMO the policy should be at least somewhat more lenient such as allowing the reviewer(s) to give an exception for this rule.

Comment (t8m) [reply policies/testing.md:15]: Ah, I see you've added this already below. OK then.

**Review by t8m — CHANGES_REQUESTED** [review]

File: policies/testing.md (hunk 7 above)
Comment (mattcaswell) [reply policies/testing.md:15]: I've only added an exception where the reviewers can agree for "difficult to reproduce error conditions". Should there be a broader exception?

Comment (t8m) [reply policies/testing.md:15]: Maybe add something like that reviewers can give an exception where writing the test would require significantly higher effort than writing the bug fix.

Comment (mattcaswell) [reply policies/testing.md:15]: I'm not sure "signficiantly higher" is quite the right bar. Its not that uncommon that writing the test is significantly more work than fixing the bug. I don't think that should necessarily be an excuse to not write a test. However, I wouldn't expect someone to have to go and write a whole new testing framework to test specific classes of things. I've made an update that says "disproportionately more effort".

**Review by t8m — APPROVED** [review]

File: policies/testing.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+ - Include files
+ - Build system
+ - Demos
+
+Pull Requests that only affect minor cosmetic output do not require tests.
+
+By agreement of the Pull Request reviewers it is acceptable for the tests to be
+added via a different Pull Request to the main Pull Request.
+
+By agreement of the Pull Request reviewers a test may be omitted where writing
```
Comment (t-j-h) [inline policies/testing.md:39]: Delete "By agreement of the Pull Request reviewers "

File: policies/testing.md [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+ - Documentation (including CHANGES.md/NEWS.md)
+ - The test suite
+ - perl utilities
+ - Include files
+ - Build system
+ - Demos
+
+Pull Requests that only affect minor cosmetic output do not require tests.
+
+By agreement of the Pull Request reviewers it is acceptable for the tests to be
```
Comment (t-j-h) [inline policies/testing.md:36]: Delete "By agreement of the Pull Request reviewers "

File: policies/testing.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+Performance testing should be performed automatically via CI on a regular basis
+for certain components as defined by the OTC from time to time.
+
+Examples of performance testing that should be considered include:
+- Individual algorithm performance operating over different input sizes
+- SSL/TLS handshake time over multiple handshakes and for different protocol
+versions and resumption/non-resumption handshakes
+
+Performance figures should be collected and tracked over time. Alerts should be
```
Comment (t-j-h) [inline policies/testing.md:54]: Remove "Alerts should be sent" - that's operational and not policy.

File: policies/testing.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+sent if performance drops below an agreed threshold versus an agreed baseline
+as set from time to time by the OTC.
```
Comment (t-j-h) [inline policies/testing.md:56]: Does not make sense to say "from time to time" ... can delete that

File: policies/testing.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+---------------
+
+This section makes recommendations that are not mandatory but should be
+considered.
+
+Pull requests that implement significant new functionality should consider
+whether fuzz tests should be added.
+
+As per the policy wording above, pull requests that implement refactoring are
+not required to add new tests. However refactoring is a good time to check that
```
Comment (t-j-h) [inline policies/testing.md:68]: It is "before refactoring" that you want to make sure the tests are decent - not during, not after.

File: policies/testing.md [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+Within this policy "functional behaviour" means what the system does, i.e.
+given a set of inputs it will produce a set of outputs. This does not include
+how the system does it. For example refactoring or performance improvements do
+not affect functional behaviour.
+
+Except where noted below:
+
+- All Pull Requests adding new functionality to the applications, libraries,
+providers or engines must include suitable tests.
+- All Pull Requests fixing a functional behavior defect in the applications,
```
Comment (t-j-h) [inline policies/testing.md:15]: behavior -> behaviour - be consistent throughout the document

File: policies/testing.md (hunk 1 above)
Comment (kroeckx) [reply policies/testing.md:16]: So any change that you can observe as caller is a function behaviour change.

File: policies/testing.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+ - The test suite
+ - perl utilities
+ - Include files
+ - Build system
+ - Demos
+
+Pull Requests that only affect minor cosmetic output do not require tests.
+
+By agreement of the Pull Request reviewers it is acceptable for the tests to be
+added via a different Pull Request to the main Pull Request.
```
Comment (kroeckx) [inline policies/testing.md:37]: Do we want to enforce that it gets written? And so that the pull requests that makes the behavioural changes can only get merged as the same time as the one testing it?

File: policies/testing.md [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+By agreement of the Pull Request reviewers a test may be omitted where writing
+the test would result in disproportionately more effort than writing the code
+being tested. For example, difficult to reproduce error conditions.
+
+Performance Testing
+-------------------
+
+Performance testing should be performed automatically via CI on a regular basis
+for certain components as defined by the OTC from time to time.
```
Comment (kroeckx) [inline policies/testing.md:47]: So that's not part of a pull request?

Comment (t8m) [reply policies/testing.md:47]: Performance testing will have to be implemented via special CI runs being run on our own HW. I am not sure we want to run these on each and every PR.

File: policies/testing.md (hunk 12 above)
Comment (mattcaswell) [reply policies/testing.md]: Added "However before refactoring..."

File: policies/testing.md (hunk 14 above)
Comment (mattcaswell) [reply policies/testing.md]: I'm not sure. The current text doesn't require it. I'd like to hear the OTC view on this.

File: policies/testing.md (hunk 15 above)
Comment (mattcaswell) [reply policies/testing.md]: Correct - not part of the pull request (I think the performance tests will be too "heavy" to run on every PR).

**Comment by mattcaswell** [comment]
Fixup applied addressing the latest feedback.

File: policies/testing.md (hunk 14 above)
Comment (t8m) [reply policies/testing.md]: **OTC does not see the need to change the policy in this way for now.**

**Comment by levitte** [comment]
Formatting suggestion:

``` diff
diff --git a/policies/testing.md b/policies/testing.md
index 13b2096..fcc8ae1 100644
--- a/policies/testing.md
+++ b/policies/testing.md
@@ -11,25 +11,27 @@ not affect functional behaviour.
 Except where noted below:
 
 - All Pull Requests adding new functionality to the applications, libraries,
-providers or engines must include suitable tests.
+  providers or engines must include suitable tests.
 - All Pull Requests fixing a functional behaviour defect in the applications,
-libraries, providers or engines must include a test for that defect.
+  libraries, providers or engines must include a test for that defect.
 
 Pull Requests that do not change the functional behaviour of the applications,
 libraries, providers or engines do not require tests to be added. For example
 the following types of changes do not require tests:
- - Changes to comments, formatting, internal naming or similar
- - Fixes for performance
- - Refactoring
+
+- Changes to comments, formatting, internal naming or similar
+- Fixes for performance
+- Refactoring
 
 To be explicit the above statement means that tests are not required for changes
 in:
- - Documentation (including CHANGES.md/NEWS.md)
- - The test suite
- - perl utilities
- - Include files
- - Build system
- - Demos
+
+- Documentation (including CHANGES.md/NEWS.md)
+- The test suite
+- perl utilities
+- Include files
+- Build system
+- Demos
 
 Pull Requests that only affect minor cosmetic output do not require tests.
 
@@ -47,9 +49,10 @@ Performance testing should be performed automatically via CI on a regular basis
 for certain components as defined by the OTC.
 
 Examples of performance testing that should be considered include:
+
 - Individual algorithm performance operating over different input sizes
 - SSL/TLS handshake time over multiple handshakes and for different protocol
-versions and resumption/non-resumption handshakes
+  versions and resumption/non-resumption handshakes
 
 Performance figures should be collected and tracked over time.
 
```

**Comment by mattcaswell** [comment]
> Formatting suggestion:

Applied.

**Comment by mattcaswell** [comment]
Vote for accepting the policy at the commit id ac9026dcba7d has started.

**Comment by mattcaswell** [comment]
Vote: [+1]

**Comment by t-j-h** [comment]
Vote: [+1]

**Comment by paulidale** [comment]
Vote: [+1]

**Comment by mattcaswell** [comment]
> While I cannot vote, I would like to ask a question: For functionality that is specific to a platform and cannot be generally tested in CI/CD systems, how will this be handled in relation to this policy?

This policy is mostly about what things should and should not have tests and has very little to do with CI. There is no distinction made about platform specific functionality. All functionality (except where the various caveats apply) should have tests (as run by "make test"). We would expect those tests to pass on all platforms. If there is no CI available for a given platform, then we won't know about test failures until someone tells us about them. In the event that there is functionality that is only available on a subset of platforms then I would expect the tests to be skipped on platforms where they are not relevant.

The platform policy is much more relevant when talking about the availability of CI (https://www.openssl.org/policies/platformpolicy.html). This tells us what platforms will have CI and which ones will not.

**Comment by slontis** [comment]
Vote: [+1]

**Comment by t8m** [comment]
Vote: [+1]

**Comment by beldmit** [comment]
Vote: [+1]

**Comment by kroeckx** [comment]
I'm voting -1 on this. I agree in principle on the direction, but I have a problem with when tests need to be added. It's my understanding that any bug fix is a functional behaviour change, and so will require a test. That sounds nice, but not all bug fixes need testing, and this policy will probably be violated on very regular basis.

**Comment by mspncp** [comment]
Vote: [+0]

I share @kroeckx's concerns that requiring tests for all bugfixes might be a little bit too strict. Do we really want a missing testcase to prevent a bugfix to get merged, even if it obviously fixes the defect?

Maybe it would be better to weaken the **must** to a **shall** in the bugfix case?

- All Pull Requests adding new functionality to the applications, libraries,
  providers or engines **must** include suitable tests.
- All Pull Requests fixing a functional behaviour defect in the applications,
  libraries, providers or engines **shall** include a test for that defect.

**Comment by romen** [comment]
Vote: [+1]

**Comment by mattcaswell** [comment]
Closing this vote now:

````
Topic: Accept openssl/technical-policies PR#13 - Testing policy as of
commit ac9026d. This will become an official OTC policy.
Proposed by: Matt Caswell
Issue link: https://github.com/openssl/technical-policies/pull/13
Public: yes
Opened: 2022-01-25
Closed: 2022-02-01
Accepted:  yes  (for: 7, against: 1, abstained: 1, not voted: 1)

  Dmitry     [+1]
  Matt       [+1]
  Pauli      [+1]
  Tim        [+1]
  Richard    [  ]
  Shane      [+1]
  Tomas      [+1]
  Kurt       [-1]
  Matthias   [+0]
  Nicola     [+1]
````

**Comment by levitte** [comment]
vote: [+0]
